File: quiz-engine/quiz-engine-dart/test/src/service/markdown_parser_test.dart
import 'package:test/test.dart';
import 'package:quiz_engine_dart/src/service/markdown_parser.dart';

void main() {
  late MarkdownParser parser;

  setUp(() => parser = MarkdownParser());

  const sampleContent = '''
# GitHub Actions Basics

## Question 1
What does CI stand for?

A) Continuous Integration
B) Code Import
C) Compile
D) Configure

**Answer:** A
**Explanation:** CI stands for Continuous Integration.

## Question 2
What is a workflow in GitHub Actions?

A) A shell script
B) An automated process triggered by events
C) A Docker image
D) A repository branch

**Answer:** B
**Explanation:** Workflows are automated processes.
''';

  group('parseContent', () {
    test('parses correct number of questions', () {
      final questions = parser.parseContent(sampleContent);
      expect(questions.length, equals(2));
    });

    test('extracts question text', () {
      final questions = parser.parseContent(sampleContent);
      expect(questions.first['question_text'],
          equals('What does CI stand for?'));
    });

    test('extracts all four options', () {
      final questions = parser.parseContent(sampleContent);
      final q = questions.first;
      expect(q['option_a'], equals('Continuous Integration'));
      expect(q['option_b'], equals('Code Import'));
      expect(q['option_c'], equals('Compile'));
      expect(q['option_d'], equals('Configure'));
    });

    test('extracts correct answer', () {
      final questions = parser.parseContent(sampleContent);
      expect(questions.first['correct_answer'], equals('A'));
    });

    test('extracts explanation', () {
      final questions = parser.parseContent(sampleContent);
      expect(questions.first['explanation'],
          equals('CI stands for Continuous Integration.'));
    });

    test('extracts section from heading', () {
      final questions = parser.parseContent(sampleContent);
      expect(questions.first['section'], equals('GitHub Actions Basics'));
    });

    test('stores sourceFile when provided', () {
      final questions = parser.parseContent(sampleContent,
          sourceFile: 'questions.md');
      expect(questions.first['source_file'], equals('questions.md'));
    });

    test('sets default usage_cycle to 1', () {
      final questions = parser.parseContent(sampleContent);
      expect(questions.first['usage_cycle'], equals(1));
    });

    test('returns empty list when no questions found', () {
      final questions = parser.parseContent('# No questions here\nJust text.');
      expect(questions, isEmpty);
    });

    test('skips block with fewer than 4 options', () {
      const content = '''
## Question 1
Incomplete question?

A) Option A
B) Option B

**Answer:** A
''';
      final questions = parser.parseContent(content);
      expect(questions, isEmpty);
    });

    test('skips block with no answer marker', () {
      const content = '''
## Question 1
No answer question?

A) Option A
B) Option B
C) Option C
D) Option D
''';
      final questions = parser.parseContent(content);
      expect(questions, isEmpty);
    });

    test('parses multiple questions in sequence', () {
      final questions = parser.parseContent(sampleContent);
      expect(questions[1]['question_text'],
          equals('What is a workflow in GitHub Actions?'));
      expect(questions[1]['correct_answer'], equals('B'));
    });

    test('handles lowercase answer letter', () {
      const content = '''
## Question 1
Question?

A) Option A
B) Option B
C) Option C
D) Option D

**Answer:** c
''';
      final questions = parser.parseContent(content);
      expect(questions.first['correct_answer'], equals('C'));
    });

    test('parses E option when present', () {
      const content = '''
## Question 1
Five option question?

A) Option A
B) Option B
C) Option C
D) Option D
E) Option E

**Answer:** E
''';
      final questions = parser.parseContent(content);
      expect(questions.first['option_e'], equals('Option E'));
      expect(questions.first['correct_answer'], equals('E'));
    });
  });
}

File: quiz-engine/quiz-engine-nodejs/test/unit/MarkdownParser.test.ts
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { parseMarkdownFile, parseMarkdownContent } from '../../src/service/MarkdownParser';

const SIMPLE_MD = `
## Q1
> What does CI stand for?
- A) Continuous Integration
- B) Code Import
- C) Compiler Install
- D) Content Index
**Answer: A**
`;

const STRUCTURED_MD = `
### Question 1 — DevOps

**Difficulty**: Easy
**Answer Type**: one
**Topic**: CI/CD

**Question**:
What does CI stand for?

- A) Continuous Integration
- B) Code Import
- C) Compiler Install
- D) Content Index

**Answer: A**
`;

const SCENARIO_MD = `
### Question 1 — Context

**Difficulty**: Medium
**Answer Type**: one
**Topic**: workflow

**Scenario**:
A developer wants to run tests on push.

**Question**:
Which trigger should they use?

- A) on: push
- B) on: schedule
- C) on: workflow_dispatch
- D) on: pull_request

**Answer: A**
`;

const WITH_EXPLANATION_MD = `
## Q1
> What does CI stand for?
- A) Continuous Integration
- B) Code Import
- C) Compiler Install
- D) Content Index
**Answer: A**
CI stands for Continuous Integration.
`;

const WITH_OPTION_E_MD = `
## Q1
> Five option question?
- A) Alpha
- B) Beta
- C) Gamma
- D) Delta
- E) Epsilon
**Answer: E**
`;

const MISSING_ANSWER_MD = `
## Q1
> No answer provided.
- A) Option A
- B) Option B
- C) Option C
- D) Option D
`;

const MISSING_OPTIONS_MD = `
## Q1
> Missing options
- A) Option A
**Answer: A**
`;

const MULTI_QUESTION_MD = `
## Q1
> First question?
- A) Option A
- B) Option B
- C) Option C
- D) Option D
**Answer: B**

## Q2
> Second question?
- A) Option A
- B) Option B
- C) Option C
- D) Option D
**Answer: C**
`;

const SOURCE_FORMAT_MD = `
## Questions

### Question 1 — DevOps

**Difficulty**: Easy
**Answer Type**: one
**Topic**: CI/CD

**Question**:
What does CI stand for?

- A) Continuous Integration
- B) Code Import
- C) Compiler Install
- D) Content Index

---

### Question 2 — DevOps

**Difficulty**: Medium
**Answer Type**: many
**Topic**: CI/CD

**Question**:
Which of these are CI tools?

- A) Jenkins
- B) Word
- C) Travis CI
- D) Paint

---

## Answer Key

| Q# | Answer(s) | Explanation | Source | Difficulty |
|----|-----------|-------------|--------|------------|
| 1 | A | CI stands for Continuous Integration. | source.md | Easy |
| 2 | A, C | Jenkins and Travis CI are CI tools. | source.md | Medium |
`;

function writeTempFile(content: string): string {
  const filePath = path.join(os.tmpdir(), `quiz-test-${Date.now()}-${Math.random()}.md`);
  fs.writeFileSync(filePath, content, 'utf-8');
  return filePath;
}

describe('MarkdownParser', () => {
  describe('parseMarkdownContent', () => {
    it('parses a simple markdown question', () => {
      const questions = parseMarkdownContent(SIMPLE_MD);
      expect(questions).toHaveLength(1);
      expect(questions[0].questionText).toBe('What does CI stand for?');
      expect(questions[0].optionA).toBe('Continuous Integration');
      expect(questions[0].correctAnswer).toBe('A');
    });

    it('parses a structured (GH-200 format) question', () => {
      const questions = parseMarkdownContent(STRUCTURED_MD);
      expect(questions).toHaveLength(1);
      expect(questions[0].questionText).toContain('What does CI stand for?');
      expect(questions[0].correctAnswer).toBe('A');
      expect(questions[0].difficulty).toBe('Easy');
      expect(questions[0].section).toBe('DevOps');
    });

    it('parses a scenario-style question', () => {
      const questions = parseMarkdownContent(SCENARIO_MD);
      expect(questions).toHaveLength(1);
      expect(questions[0].questionText).toContain('Which trigger should they use?');
      expect(questions[0].correctAnswer).toBe('A');
    });

    it('parses explanation text after answer', () => {
      const questions = parseMarkdownContent(WITH_EXPLANATION_MD);
      expect(questions).toHaveLength(1);
      expect(questions[0].explanation).toContain('CI stands for Continuous Integration');
    });

    it('parses 5-option question', () => {
      const questions = parseMarkdownContent(WITH_OPTION_E_MD);
      expect(questions).toHaveLength(1);
      expect(questions[0].optionE).toBe('Epsilon');
      expect(questions[0].correctAnswer).toBe('E');
    });

    it('parses multiple questions from one file', () => {
      const questions = parseMarkdownContent(MULTI_QUESTION_MD);
      expect(questions).toHaveLength(2);
      expect(questions[0].correctAnswer).toBe('B');
      expect(questions[1].correctAnswer).toBe('C');
    });

    it('throws on missing answer line', () => {
      expect(() => parseMarkdownContent(MISSING_ANSWER_MD)).toThrow();
    });

    it('throws on missing required options', () => {
      expect(() => parseMarkdownContent(MISSING_OPTIONS_MD)).toThrow();
    });

    it('returns empty array for empty content', () => {
      const questions = parseMarkdownContent('');
      expect(questions).toHaveLength(0);
    });

    it('returns empty array for content with no question headers', () => {
      const questions = parseMarkdownContent('# Just a heading\nSome text here.');
      expect(questions).toHaveLength(0);
    });
  });

  describe('parseMarkdownFile', () => {
    it('parses a valid markdown file', () => {
      const filePath = writeTempFile(SIMPLE_MD);
      try {
        const questions = parseMarkdownFile(filePath);
        expect(questions).toHaveLength(1);
        expect(questions[0].correctAnswer).toBe('A');
      } finally {
        fs.unlinkSync(filePath);
      }
    });

    it('throws on missing file', () => {
      expect(() => parseMarkdownFile('/nonexistent/path/file.md')).toThrow();
    });

    it('parses file with multiple questions', () => {
      const filePath = writeTempFile(MULTI_QUESTION_MD);
      try {
        const questions = parseMarkdownFile(filePath);
        expect(questions).toHaveLength(2);
      } finally {
        fs.unlinkSync(filePath);
      }
    });
  });

  describe('source format with answer key table', () => {
    it('parses single-answer questions from answer key', () => {
      const questions = parseMarkdownContent(SOURCE_FORMAT_MD);
      expect(questions).toHaveLength(1);
      expect(questions[0].correctAnswer).toBe('A');
    });

    it('extracts explanation from answer key', () => {
      const questions = parseMarkdownContent(SOURCE_FORMAT_MD);
      expect(questions[0].explanation).toBe('CI stands for Continuous Integration.');
    });

    it('extracts section from source format header', () => {
      const questions = parseMarkdownContent(SOURCE_FORMAT_MD);
      expect(questions[0].section).toBe('DevOps');
    });

    it('skips many-type questions in source format', () => {
      const questions = parseMarkdownContent(SOURCE_FORMAT_MD);
      const texts = questions.map((q) => q.questionText);
      expect(texts.some((t) => t.includes('CI tools'))).toBe(false);
    });
  });
});
